File: 0402-remove-k-digits/solution.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        stack = []
        for digit in num:
            while stack and k > 0 and stack[-1] > digit:
                stack.pop()
                k -= 1
            stack.append(digit)

        # Remove digits from the end if k is still positive
        while k > 0 and stack:
            stack.pop()
            k -= 1

        # Remove leading zeros
        while stack and stack[0] == '0':
            stack.pop(0)

        # If stack is empty, return "0"
        if not stack:
            return "0"

        # Join digits in the stack and return the resulting string
        return ''.join(stack)

        
    
        

File: 0402-remove-k-digits/solution.java
public class Solution {
    public String removeKdigits(String num, int k) {
        Deque<Character> deque = new LinkedList<>();
        for (char digit : num.toCharArray()) {
            while (!deque.isEmpty() && k > 0 && deque.peekLast() > digit) {
                deque.pollLast();
                k--;
            }
            deque.offerLast(digit);
        }

        // Remove digits from the end if k is still positive
        while (k > 0 && !deque.isEmpty()) {
            deque.pollLast();
            k--;
        }

        // Remove leading zeros
        while (!deque.isEmpty() && deque.peek() == '0') {
            deque.pollFirst();
        }

        // If deque is empty, return "0"
        if (deque.isEmpty()) {
            return "0";
        }

        // Build the resulting string from deque
        StringBuilder result = new StringBuilder();
        for (char c : deque) {
            result.append(c);
        }
        return result.toString();
    }
}

